docassemble/ALToolbox/al_income.py

```diff
@@ -0,0 +1,1015 @@
+            times_per_year
+        )
+
+
+class ALJobList(ALIncomeList):
+    """
+    Represents a list of ALJobs. Adds the `.gross_total()` and
+    `.net_total()` methods to the ALIncomeList class. It's a more common
+    way of reporting income than ALItemizedJobList.
+    """
+
+    def init(self, *pargs, **kwargs):
+        super().init(*pargs, **kwargs)
+        self.object_type = ALJob
+
+    def total(self, times_per_year=1, source=None):
+        """Alias for self.gross_total() to integrate with ALIncomeList math."""
+        return self.gross_total(times_per_year=times_per_year, source=source)
+
+    def gross_total(self, times_per_year=1, source=None):
+        """
+        Returns the sum of the gross incomes of its ALJobs divided by the time
+        times_per_year. You can filter the jobs by `source`. `source` can be a
+        string or a list.
+
+        `times_per_year` is some demoninator of a year. E.g, to express a weekly
+        period, use 52. The default is 1 (a year).
+        """
+        self._trigger_gather()
+        result = Decimal(0)
+        if times_per_year == 0:
+            return result
+        if source is None:
+            for job in self.elements:
+                result += Decimal(job.gross_total(times_per_year=times_per_year))
+        elif isinstance(source, list):
+            for job in self.elements:
+                if job.source in source:
+                    result += Decimal(job.gross_total(times_per_year=times_per_year))
+        else:
+            for job in self.elements:
+                if job.source == source:
+                    result += Decimal(job.gross_total(times_per_year=times_per_year))
+        return result
+
+    def net_total(self, times_per_year=1, source=None):
+        """
+        Returns the sum of the net incomes of its ALJobs divided by the time
+        times_per_year. You can filter the jobs by `source`. `source` can be a
+        string or a list. Only returns a correct total if the job is non-hourly.
+        Leaving out `source` will use all sources.
+
+        `times_per_year` is some demoninator of a year. E.g, to express a weekly
+        period, use 52. The default is 1 (a year).
+        """
+        self._trigger_gather()
+        result = Decimal(0)
+        if times_per_year == 0:
+            return result
+        if source is None:
+            for job in self.elements:
+                result += Decimal(job.net_total(times_per_year=times_per_year))
+        elif isinstance(source, list):
+            for job in self.elements:
+                if job.source in source:
+                    result += Decimal(job.net_total(times_per_year=times_per_year))
+        else:
+            for job in self.elements:
+                if job.source == source:
+                    result += Decimal(job.net_total(times_per_year=times_per_year))
+        return result
+
+
+class ALAsset(ALIncome):
+    """Like ALIncome but the `value` attribute is optional."""
```

Comment:
Does this mean an `ALAsset` should have a `.market_value` instead of a `.value` and use that in its `.total()` method? Do there need to be other changes too? For example, for the `.balance`?

Reply:
An asset can have both a value, which represents income, and a market value, which represents the asset's current worth.

E.g., if you have a checking account with a $1,000 balance that earns 1% interest each year:

- value is 10.0
- times_per_year would be 1.0
- market_value would be 1000.0

A financial statement would want you to list the two things separately. Usually there are two separate limits for income and assets. E.g., for SSI you can earn up to ~ $800/month and still receive something, but you're only allowed to have up to $2,000 in assets.

The total() of an IncomeList reflects the total income, not the total value of the items. You need to report both income from wages and income earned as interest on your savings and checking accounts, no matter how small those figures might be. So these still need to be two different methods.

Reply:
Some assets, such as houses and vehicles, also have an associated loan. that's what the `balance` attribute is for. E.g., consider the example for a house that was owned by a landlord:

- value of 1500.0 (rent income)
- times_per_year of 12.0 (rent is paid monthly)
- market_value of 100000.0 (the current estimated sales price)
- balance of 80000.0 (mortgage)

For most people, houses and vehicles don't earn any income so the value is likely to be zero, but sometimes they do earn income.

Reply:
Thanks for the explanation. Since ALAsset itself only deals with `.value` and never refers to `.market_value`, I'm not sure what to document about it. To be honest, I'm not sure what to document about `.market_value` - just that ALAsset can have one, though only the ALAssetList does anything with it? Also, if we're documenting `.market_value`, do we also need to document `.balance`?

Reply:
Yes, both should be noted as the intent of the ALAsset is to provide one or both of those attributes. You can't put them in the class definition because that interferes with the way Docassemble resolves undefined attributes, but they should be mentioned in the docstring.